ext/openssl/openssl.c

```diff
@@ -3821,49 +3987,96 @@ static int php_openssl_is_private_key(EVP_PKEY* pkey)
 }
 /* }}} */
 
-#define OPENSSL_GET_BN(_array, _bn, _name) do {						\
-		if (_bn != NULL) {							\
-			int len = BN_num_bytes(_bn);					\
-			zend_string *str = zend_string_alloc(len, 0);			\
-			BN_bn2bin(_bn, (unsigned char*)ZSTR_VAL(str));			\
-			ZSTR_VAL(str)[len] = 0;						\
-			add_assoc_str(&_array, #_name, str);				\
-		}									\
+#define OPENSSL_GET_BN(_array, _bn, _name) do { \
+		if (_bn != NULL) { \
+			int len = BN_num_bytes(_bn); \
+			zend_string *str = zend_string_alloc(len, 0); \
+			BN_bn2bin(_bn, (unsigned char*)ZSTR_VAL(str)); \
+			ZSTR_VAL(str)[len] = 0; \
+			add_assoc_str(&_array, #_name, str); \
+		} \
 	} while (0);
 
-#define OPENSSL_PKEY_GET_BN(_type, _name) do {						\
-		if (pkey->pkey._type->_name != NULL) {					\
-			OPENSSL_GET_BN(_type, pkey->pkey._type->_name, _name);		\
-		}									\
-	} while (0);
+#define OPENSSL_PKEY_GET_BN(_type, _name) OPENSSL_GET_BN(_type, _name, _name)
 
-#define OPENSSL_PKEY_SET_BN(_ht, _type, _name) do {					\
-		zval *bn;								\
-		if ((bn = zend_hash_str_find(_ht, #_name, sizeof(#_name)-1)) != NULL && \
-				Z_TYPE_P(bn) == IS_STRING) {				\
-			_type->_name = BN_bin2bn(					\
-				(unsigned char*)Z_STRVAL_P(bn),				\
-				(int)Z_STRLEN_P(bn), NULL);				\
-		}										\
+#define OPENSSL_PKEY_SET_BN(_data, _name) do { \
+		zval *bn; \
+		if ((bn = zend_hash_str_find(Z_ARRVAL_P(_data), #_name, sizeof(#_name)-1)) != NULL && \
+				Z_TYPE_P(bn) == IS_STRING) { \
```

Comment:
Not related to this PR, but this should probably include a DEREF in case the data array contains references. (Unless this is not user-supplied, I did not check.)